### PR TITLE
Update: skip runtime status copy after successful execution

### LINKED ISSUE #1408: [Code Health] Clarify or retire packed graph-output copy-back contract
### Category

Technical Debt (cleanup, refactor)

### Component

Host Runtime

### Description

The PTO2 shared-memory header still defines `graph_output_ptr` and
`graph_output_size` as an orchestrator-to-host packed-output contract.
Host finalization reads these fields and, when non-zero, copies the first
output tensor from the packed address instead of its normal device buffer.

The current tree initializes both fields to zero but has no in-tree writer
that publishes a non-zero packed output. The comments and runtime logic docs
still describe the protocol as supported.

This stale contract couples two unrelated responsibilities to the same
successful-run D2H copy:

- fatal orchestration and scheduler diagnostics;
- successful packed-output routing.

The ambiguity blocks a safe optimization. Skipping the status-header D2H
after a successful execution also leaves the graph-output fields at their
zero-initialized host values, silently disabling the documented packed-output
path. Keeping the unconditional copy preserves a protocol that appears unused
and adds host/device synchronization overhead to every successful run.

We should explicitly decide whether packed graph-output copy-back remains a
supported contract or is legacy dead code.

### Location

- `src/a2a3/runtime/tensormap_and_ringbuffer/host/runtime_maker.cpp:970-1045`
- `src/a5/runtime/tensormap_and_ringbuffer/host/runtime_maker.cpp:925-1000`
- `src/a2a3/runtime/host_build_graph/host/runtime_maker.cpp:975-1032`
- `src/a2a3/runtime/tensormap_and_ringbuffer/runtime/pto_shared_memory.h:132-135`
- `src/a5/runtime/tensormap_and_ringbuffer/runtime/pto_shared_memory.h:132-135`
- `src/a2a3/runtime/host_build_graph/runtime/pto_shared_memory.h:126-129`
- `src/*/runtime/*/docs/RUNTIME_LOGIC.md`, shared-header field tables

### Proposed Fix

First audit whether any supported external or historical orchestration still
publishes these fields.

If the protocol is obsolete:

- remove the host packed-output fallback;
- retain or mark the header slots as reserved if layout compatibility matters;
- update the shared-header comments and runtime logic documentation;
- then read the device status header only when execution or stream sync fails.

If the protocol is still supported:

- separate successful output metadata from fatal-status diagnostics, or add an
  explicit host-visible capability indicating that packed output is enabled;
- preserve the packed-output read only for that capability;
- add a regression test with a non-zero packed-output pointer.

In either case, add tests proving that normal successful runs avoid the status
D2H while failed runs retain detailed device error reporting.

### Priority

Medium (minor risk, should fix in next few releases)

### PR DESCRIPTION
**Pull request by zmnobug**
## Summary

- propagate execution results into runtime finalization across sim and onboard
- skip the PTO2 shared-header D2H copy after successful execution
- never copy tensors back after a failed execution, even when no device status was published
- prefer detailed device runtime status over generic runner or stream errors
- retire the unused packed graph-output fallback while preserving shared-memory layout compatibility

## Testing

- [x] .venv/bin/pip install --no-build-isolation -e . (a2a3sim, a5sim, a2a3, and a5 runtimes)
- [x] cmake --build tests/ut/cpp/build --target test_trb_runtime_temp_buffer -j 8
- [x] ctest --test-dir tests/ut/cpp/build -R test_trb_runtime_temp_buffer --output-on-failure (10 tests passed)
- [x] ctest --test-dir tests/ut/cpp/build -LE requires_hardware --output-on-failure (59 passed in sandbox; socket test passed separately outside the sandbox)
- [x] clang-tidy plus header, English-only, formatting, and diff checks
- [x] GitHub CI full matrix, including st-onboard-a2a3 and st-onboard-a5

## Additional Context

The legacy packed graph-output fields remain in the shared-memory header only for layout compatibility. Runtime finalization no longer reads them and copies each output from its tensor allocation.

Fixes #1408

### CODE REVIEW

File: src/a2a3/runtime/host_build_graph/host/runtime_maker.cpp [hunk 1]
```diff
@@ -977,7 +978,9 @@ extern "C" int validate_runtime_impl(Runtime *runtime, const HostApi *api) {
     PTO2SharedMemoryHeader host_header;
     memset(&host_header, 0, sizeof(host_header));
 
-    runtime_status = pto2_read_runtime_status(runtime, api, &host_header);
+    if (execution_rc != 0) {
+        runtime_status = pto2_read_runtime_status(runtime, api, &host_header);
+    }
```
[the remark this answers is not in the document]
Comment (zmnobug) [reply src/a2a3/runtime/host_build_graph/host/runtime_maker.cpp:980]: Fixed in 343b45ae — failed execution now sets the copy-back skip condition before reading the diagnostic header, so a zero or unavailable device status cannot trigger tensor D2H.

File: src/common/platform/onboard/host/c_api_shared.cpp [hunk 2]
```diff
@@ -605,13 +605,13 @@ int simpler_run(
             rc = runner->run(*r, *config);
         }
         if (rc != 0) {
-            validate_runtime_impl(r, &g_host_api);
+            validate_runtime_impl(r, &g_host_api, rc);
             return rc;
         }
```
[the remark this answers is not in the document]
Comment (zmnobug) [reply src/common/platform/onboard/host/c_api_shared.cpp]: Fixed in 343b45ae — sim and onboard failure paths now prefer the detailed status returned by runtime validation and fall back to the original bind/runner error when validation has no device status.

File: src/a5/runtime/host_build_graph/host/runtime_maker.cpp [hunk 3]
```diff
@@ -466,10 +466,11 @@ int bind_callable_to_runtime_impl(
  * 2. Frees device memory for recorded tensors
  * 3. Clears tensor pair state
  *
- * @param runtime  Pointer to Runtime
+ * @param runtime       Pointer to Runtime
+ * @param execution_rc  Status returned by DeviceRunner::run
  * @return 0 on success, -1 on failure
  */
-int validate_runtime_impl(Runtime *runtime, const HostApi *api) {
+int validate_runtime_impl(Runtime *runtime, const HostApi *api, int /* execution_rc */) {
```
[the remark this answers is not in the document]
Comment (zmnobug) [reply src/a5/runtime/host_build_graph/host/runtime_maker.cpp]: Fixed in 343b45ae — a5 host_build_graph now consumes execution_rc and guards the tensor copy-back loop, while still running cleanup after failed execution.

File: src/a2a3/runtime/tensormap_and_ringbuffer/host/runtime_maker.cpp [hunk 4]
```diff
@@ -972,7 +973,9 @@ extern "C" int validate_runtime_impl(Runtime *runtime, const HostApi *api) {
     PTO2SharedMemoryHeader host_header;
     memset(&host_header, 0, sizeof(host_header));
 
-    runtime_status = pto2_read_runtime_status(runtime, api, &host_header);
+    if (execution_rc != 0) {
+        runtime_status = pto2_read_runtime_status(runtime, api, &host_header);
+    }
```
[the remark this answers is not in the document]
Comment (zmnobug) [reply src/a2a3/runtime/tensormap_and_ringbuffer/host/runtime_maker.cpp:975]: Fixed in 343b45ae — removed the unused packed graph-output fallback from all three affected finalizers. The fields remain reserved only for shared-memory layout compatibility, the runtime docs now state that contract, and the success regression test verifies tensor-local copy-back.

[the remark this answers is not in the document]
Comment (zmnobug) [reply src/a2a3/runtime/tensormap_and_ringbuffer/host/runtime_maker.cpp:975]: Fixed in 343b45ae — failed execution now unconditionally skips tensor copy-back; the header read is diagnostic-only and a zero status no longer re-enables output copies.

File: src/a5/runtime/tensormap_and_ringbuffer/host/runtime_maker.cpp [hunk 5]
```diff
@@ -927,7 +928,9 @@ extern "C" int validate_runtime_impl(Runtime *runtime, const HostApi *api) {
     PTO2SharedMemoryHeader host_header;
     memset(&host_header, 0, sizeof(host_header));
 
-    runtime_status = pto2_read_runtime_status(runtime, api, &host_header);
+    if (execution_rc != 0) {
+        runtime_status = pto2_read_runtime_status(runtime, api, &host_header);
+    }
```
[the remark this answers is not in the document]
Comment (zmnobug) [reply src/a5/runtime/tensormap_and_ringbuffer/host/runtime_maker.cpp:930]: Fixed in 343b45ae — failed execution now unconditionally skips tensor copy-back; the header read is diagnostic-only and a zero status no longer re-enables output copies.

**Review by ChaoZheng109 — COMMENTED** [review]
Overall this is a clean, well-scoped implementation of #1408 and CI (incl. st-onboard-a2a3/a5) is green. One nit: the skip-copy-back `LOG_WARN` now reprints `execution_rc`, which duplicates an error code that is already reported centrally. See the inline comments — same change applies to all four `runtime_maker.cpp`.

File: src/a2a3/runtime/host_build_graph/host/runtime_maker.cpp [hunk 6]
```diff
@@ -969,32 +970,23 @@ extern "C" int validate_runtime_impl(Runtime *runtime, const HostApi *api) {
...
-        graph_out_ptr = host_header.graph_output_ptr;
-        graph_out_size = host_header.graph_output_size;
-        if (graph_out_ptr != 0) {
-            LOG_INFO_V0("Graph output buffer: ptr=0x%" PRIx64 ", size=%" PRIu64, graph_out_ptr, graph_out_size);
-        }
     }
 
     if (skip_tensor_copy_back) {
-        LOG_WARN("Skipping tensor copy-back because PTO2 runtime reported fatal status");
+        LOG_WARN("Skipping tensor copy-back because execution failed (rc=%d)", execution_rc);
```
Comment (ChaoZheng109) [inline src/a2a3/runtime/host_build_graph/host/runtime_maker.cpp:988]: This WARN previously carried no code (`"...PTO2 runtime reported fatal status"`); the PR added `(rc=%d), execution_rc`. But this rc is already surfaced centrally on the failure path — `aclrtSynchronizeStreamWithTimeout (AICPU) failed: %d` (device_runner_base.cpp:1229), `LOG_RUNTIME_FAILURE(...)`, and the final `run failed with code X` exception. Reprinting it here scatters the same error code across several log lines. This WARN only needs to state the *action* (copy-back was skipped), not restate the code:

```cpp
LOG_WARN("Skipping tensor copy-back because execution failed");
```

Same change applies to the other three runtime makers (see companion comments).

File: src/a2a3/runtime/tensormap_and_ringbuffer/host/runtime_maker.cpp [hunk 7]
```diff
@@ -995,19 +995,11 @@ extern "C" int validate_runtime_impl(Runtime *runtime, const HostApi *api) {
...
-        graph_out_ptr = host_header.graph_output_ptr;
-        graph_out_size = host_header.graph_output_size;
-        if (graph_out_ptr != 0) {
-            LOG_INFO_V0("Graph output buffer: ptr=0x%" PRIx64 ", size=%" PRIu64, graph_out_ptr, graph_out_size);
-        }
     }
 
     if (skip_tensor_copy_back) {
-        LOG_WARN("Skipping tensor copy-back because PTO2 runtime reported fatal status");
+        LOG_WARN("Skipping tensor copy-back because execution failed (rc=%d)", execution_rc);
```
Comment (ChaoZheng109) [inline src/a2a3/runtime/tensormap_and_ringbuffer/host/runtime_maker.cpp:1001]: Same as in the a2a3 host_build_graph maker: drop `(rc=%d), execution_rc` — the code is already reported centrally (stream-sync WARN, `LOG_RUNTIME_FAILURE`, and the `run failed with code` exception). Keep just `LOG_WARN("Skipping tensor copy-back because execution failed");`.

File: src/a5/runtime/tensormap_and_ringbuffer/host/runtime_maker.cpp [hunk 8]
```diff
@@ -950,19 +950,11 @@ extern "C" int validate_runtime_impl(Runtime *runtime, const HostApi *api) {
...
-        graph_out_ptr = host_header.graph_output_ptr;
-        graph_out_size = host_header.graph_output_size;
-        if (graph_out_ptr != 0) {
-            LOG_INFO_V0("Graph output buffer: ptr=0x%" PRIx64 ", size=%" PRIu64, graph_out_ptr, graph_out_size);
-        }
     }
 
     if (skip_tensor_copy_back) {
-        LOG_WARN("Skipping tensor copy-back because PTO2 runtime reported fatal status");
+        LOG_WARN("Skipping tensor copy-back because execution failed (rc=%d)", execution_rc);
```
Comment (ChaoZheng109) [inline src/a5/runtime/tensormap_and_ringbuffer/host/runtime_maker.cpp:956]: Same nit: drop the `(rc=%d), execution_rc` from this skip WARN so the error code isn't duplicated across log lines.

File: src/a5/runtime/host_build_graph/host/runtime_maker.cpp [hunk 9]
```diff
@@ -487,15 +488,19 @@ int validate_runtime_impl(Runtime *runtime, const HostApi *api) {
...
-        const TensorPair &pair = tensor_pairs[i];
-        int copy_rc = api->copy_from_device(pair.host_ptr, pair.dev_ptr, pair.size);
-        if (copy_rc != 0) {
-            LOG_ERROR("Failed to copy tensor %d from device: %d", i, copy_rc);
-            rc = copy_rc;
-            // Continue with cleanup anyway
-        } else {
-            LOG_DEBUG("Tensor %d: %zu bytes copied to host", i, pair.size);
+    if (execution_rc != 0) {
+        LOG_WARN("Skipping tensor copy-back because execution failed (rc=%d)", execution_rc);
```
Comment (ChaoZheng109) [inline src/a5/runtime/host_build_graph/host/runtime_maker.cpp:492]: This one is newly introduced by the PR. Same nit — it only needs to state that copy-back was skipped; the rc is reported centrally elsewhere, so drop `(rc=%d), execution_rc`.

**Review by ChaoZheng109 — APPROVED** [review]

**Comment by ChaoZheng109** [comment]
### Follow-up cleanup PR (deferred from this PR)

This PR intentionally kept `graph_output_ptr` / `graph_output_size` as
reserved-zero padding to keep its blast radius on the perf optimization
(skip the success-path status D2H). "Reserved forever" shouldn't be the
end state, so I'll open a dedicated follow-up so each PR has one clear
purpose. Scope:

1. **Delete the two reserved fields for real.** Remove `graph_output_ptr`
   / `graph_output_size` from the 3 private `pto_shared_memory.h` copies
   (a2a3 host_build_graph, a2a3/a5 tensormap_and_ringbuffer), drop their
   zero-init `.store(0, …)` in the matching `pto_shared_memory.cpp`, and
   delete the "Reserved" rows from the 3 `RUNTIME_LOGIC.md` tables
   (instead of documenting permanent padding). They have no writer and,
   after this PR, no reader.

2. **Add a layout guard so a future drift fails loudly, not silently.**
   The header has no runtime version/magic field today, so any missed
   lockstep rebuild between the independently-compiled host `.so` / AICPU
   `.so` / AICore / orchestration `.so` is a *silent* ABI mismatch (all
   access is compile-time `offsetof`). Add a
   `static_assert(sizeof(PTO2SharedMemoryHeader) == …)` and/or a
   `version` field validated at load. This is the piece that makes it
   safe to shrink the struct at all.

3. **Force a lockstep rebuild.** Clear the `build/lib` prebuilt-runtime
   cache so runtime binaries and JIT-compiled orchestration `.so` are
   rebuilt against the new layout together (the project has already been
   bitten by `build/lib` lagging source — cf. the `pto_isa.pin` guards).

4. **Finish the deferred log cleanup.** The skip-copy-back
   `LOG_WARN("…(rc=%d)", execution_rc)` added here (4× `runtime_maker.cpp`)
   reprints an error code already reported centrally (stream-sync WARN,
   `LOG_RUNTIME_FAILURE`, and the `run failed with code` exception). Drop
   the `(rc=%d)`; the WARN only needs to state that copy-back was skipped.

5. **(Optional) Surface both codes on failure instead of swapping.** This
   PR makes `simpler_run` prefer the PTO2 device code (e.g. `-100`) over
   the ACL code (e.g. `507046`) in the exception. Both are useful — the
   ACL code names the host mechanism that fired, the PTO2 code names the
   device-side cause. Consider carrying both, e.g.
   `run failed with code 507046 (device: -100 scheduler-timeout)`, rather
   than replacing one with the other.

Rationale for splitting: items 1–3 touch an offset-sensitive shared-memory
ABI and need a coordinated rebuild + a layout guard; bundling them into the
perf PR would have widened the blast radius for zero functional gain.
Keeping them here as a tracked follow-up is the clean separation.
